### ISSUE TITLE
HTTP requests are taking a long time to process

### ISSUE THREAD

**Issue by Half-Shot** [matrix-org/synapse#2910]
### Description

Synapse is heaving and seems to be having trouble even doing the simplest of operations without taking a long time to perform them. Incremental syncs time out often, and even talking to the versions endpoint seems to be taking 10s of seconds. This is not behind a load balancer.

I have tried clearing extremities in the past, but obviously this issue is much deeper than that.

The system is not running, nor anywhere close to full load. The synapse master process is often sitting at 1.5 load on it's own.

### Steps to reproduce

```
Halfy-VM ~ » time curl localhost:8008/_matrix/client/versions
{
    "versions": [
        "r0.0.1",
        "r0.1.0",
        "r0.2.0",
        "r0.3.0"
    ]
}
curl localhost:8008/_matrix/client/versions  0.00s user 0.00s system 0% cpu 12.593 total
```
homeserver.log doesn't seem to have anything of use, but I will provide it if it helps.

[homeserver.txt](https://github.com/matrix-org/synapse/files/1755446/homeserver.txt)

### Version information

- **Homeserver**: matrix.half-shot.uk

- **Version**:        Synapse/0.26.0
- **Install method**: offical deb package
- **Platform**:       straight forward 16.04 ubuntu install, no docker or the like. Behind nginx but not used in this example.

**Comment by ara4n** [comment]
we won't be able to help with this without logs - please can you DM them to me?

**Comment by Half-Shot** [comment]
Done

**Comment by ara4n** [comment]
So grepping the logs for /versions, synapse thinks it's generating the response in  a few ms:

```
homeserver.log.1:2018-02-24 19:00:50,479 - synapse.access.http.8008 - 91 - INFO - GET-130522- - - 8008 - {None} Processed request: 0ms (0ms, 0ms) (0ms/0) 99B 200 "GET /_matrix/client/versions HTTP/1.0" "curl/7.47.0"
homeserver.log.1:2018-02-24 19:00:52,668 - synapse.access.http.8008 - 91 - INFO - GET-130532- - - 8008 - {None} Processed request: 0ms (4ms, 0ms) (0ms/0) 99B 200 "GET /_matrix/client/versions HTTP/1.0" "curl/7.47.0"
```

So if it's taking 12s to return, either your client is somehow blocked on DNS - or possibly the twisted reactor is completely overloaded and it's taking 12s to write back to you with the missing data.

There are a handful of disproportionately long requests, e.g...

```
2018-02-24 22:18:08,920 - synapse.access.http.8008 - 91 - INFO - POST-151921- - - 8008 - {@Half-Shot:half-shot.uk} Processed request: 5588ms (4ms, 0ms) (131ms/4) 2B 200 "POST /_matrix/client/r0/pushers/set?access_token=<redacted> HTTP/1.0" "Riot.im/0.8.1 (Linux; U; Android 8.0.0; ONEPLUS A5000 Build/OPR6.170623.01; Flavour GooglePlay; MatrixAndroidSDK 0.9.0)"
```

...but nothing too dramatic.

So, unsure what's going on here.  Are you leaking FDs? What does your lsof look like for the process?

**Comment by Half-Shot** [comment]
DNS seems unlikely, it's just talking to localhost over CURL which is about as simple as I can make it.
Doesn't look like a FD leak, currently on 192 FDs :|.

I'm starting to think it might be twisted suffering, but I'm not sure how I could trace that?

**Comment by ara4n** [comment]
can you get a pyflame out of it whilst it’s trying to process the /versions?

**Comment by ara4n** [comment]
oops. also can you try correlating the timings of the beginning and end of the curl with the timings of the received or processed loglines in the synapse logs? ie is it taking a while to receive the request or send the response?

**Comment by Half-Shot** [comment]
On it :+1: . Just did two trial runs and got 11.8 seconds and 4.8 seconds for the versions endpoint back to back :/

**Comment by Half-Shot** [comment]
Did the best I could, but it was erratic where sometimes it was returning near instantly but the median time was about 4 seconds, sometimes leaping up to 12..15.

### Curl
```
Halfy-VM ~ » echo `date +%T:%N` && curl -v localhost:8008/_matrix/client/versions && echo `date +%T:%N`   
11:54:17:412341439
*   Trying ::1...
* connect to ::1 port 8008 failed: Connection refused
*   Trying 127.0.0.1...
* Connected to localhost (127.0.0.1) port 8008 (#0)
> GET /_matrix/client/versions HTTP/1.1
> Host: localhost:8008
> User-Agent: curl/7.47.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Length: 99
< Access-Control-Allow-Headers: Origin, X-Requested-With, Content-Type, Accept, Authorization
< Server: Synapse/0.26.0
< Date: Sun, 25 Feb 2018 10:54:17 GMT
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, POST, PUT, DELETE, OPTIONS
< Content-Type: application/json
< 
{
    "versions": [
        "r0.0.1",
        "r0.1.0",
        "r0.2.0",
        "r0.3.0"
    ]
}
* Connection #0 to host localhost left intact
11:54:22:019900163
```

### PyFlame
``
rm -f prof.txt && echo `date +%T:%N` && sudo ./pyflame -s 15 -r 0.01 -p 17857 -o prof.txt && echo `date +%T:%N`
11:54:16:789304644
11:54:31:818510262
``
[prof.txt](https://github.com/matrix-org/synapse/files/1755912/prof.txt)

### Synapse log lines

```
2018-02-25 11:54:17,805 - synapse.access.http.8008 - 59 - INFO - GET-233082- - - 8008 - Received request: GET /_matrix/client/versions
2018-02-25 11:54:17,829 - synapse.access.http.8008 - 91 - INFO - GET-233082- - - 8008 - {None} Processed request: 23ms (8ms, 0ms) (0ms/0) 99B 200 "GET /_matrix/client/versions HTTP/1.1" "curl/7.47.0"
2018-02-25 11:54:17,830 - synapse.util.async - 237 - INFO - PUT-233081- Releasing linearizer lock 'fed_txn_handler' for key ('unrelenting.technology', u'1518963833642')
2018-02-25 11:54:17,831 - synapse.access.http.8444 - 91 - INFO - PUT-233081- - - 8444 - {unrelenting.technology} Processed request: 522ms (28ms, 3ms) (27ms/2) 11B 200 "PUT /_matrix/federation/v1/send/1518963833642/ HTTP/1.0" "Synapse/0.26.0"
2018-02-25 11:54:20,979 - synapse.state - 502 - INFO - PUT-233067- Asking for 8484 conflicted events
2018-02-25 11:54:21,270 - synapse.state - 519 - INFO - PUT-233067- Asking for 4 auth events
2018-02-25 11:54:21,822 - synapse.http.matrixfederationclient - 212 - WARNING - - {PUT-O-847001} Sending request failed to corzntin.fr: PUT matrix://corzntin.fr/_matrix/federation/v1/send/1519425659089/: TimeoutError('',)
2018-02-25 11:54:21,829 - synapse.http.endpoint - 264 - INFO - - Connecting to 88.85.138.176:34421
2018-02-25 11:54:21,833 - synapse.handlers.presence - 327 - INFO - - Handling presence timeouts
2018-02-25 11:54:21,835 - synapse.handlers.typing - 78 - INFO - - Checking for typing timeouts
2018-02-25 11:54:21,927 - synapse.metrics - 162 - INFO - - Collecting gc 0
2018-02-25 11:54:22,021 - synapse.access.http.8444 - 59 - INFO - PUT-233083- - - 8444 - Received request: PUT /_matrix/federation/v1/send/1519520895297/
2018-02-25 11:54:22,044 - synapse.federation.transport.server - 140 - INFO - PUT-233083- Request from pdxmesh.net
2018-02-25 11:54:22,060 - synapse.federation.transport.server - 251 - INFO - PUT-233083- Received txn 1519520895297 from pdxmesh.net. (PDUs: 0, EDUs: 1, failures: 0)
2018-02-25 11:54:22,061 - synapse.util.async - 230 - INFO - PUT-233083- Acquired uncontended linearizer lock 'fed_txn_handler' for key ('pdxmesh.net', u'1519520895297')
2018-02-25 11:54:22,072 - synapse.http.endpoint - 264 - INFO - - Connecting to matrix.dmp-develop.com:8448
2018-02-25 11:54:22,130 - synapse.http.endpoint - 264 - INFO - - Connecting to molokola.com:8448
2018-02-25 11:54:22,146 - synapse.http.endpoint - 264 - INFO - - Connecting to susi.praha.bahnhof.cz:8448
2018-02-25 11:54:22,733 - synapse.util.async - 230 - INFO - GET-233078- Acquired uncontended linearizer lock 'state_resolve_lock' for key frozenset([2907349, 2908751, 2894253L, 2908591])
2018-02-25 11:54:22,734 - synapse.state - 323 - INFO - GET-233078- Resolving state for !cURbafjkfsMDVwdRDQ:matrix.org with 4 groups
2018-02-25 11:54:22,850 - synapse.state - 342 - INFO - GET-233078- Resolving conflicted state for u'!cURbafjkfsMDVwdRDQ:matrix.org'
2018-02-25 11:54:22,902 - synapse.state - 502 - INFO - GET-233078- Asking for 1696 conflicted events
2018-02-25 11:54:22,927 - synapse.state - 519 - INFO - GET-233078- Asking for 8 auth events
2018-02-25 11:54:23,111 - synapse.util.async - 237 - INFO - GET-233078- Releasing linearizer lock 'state_resolve_lock' for key frozenset([2907349, 2908751, 2894253L, 2908591])
```

**Comment by Half-Shot** [comment]
Oh and one more thing if it helps: Curl blocks waiting for a reply rather than the body, i.e we don't receive the headers and then wait for the body, it all comes down in one chunk a while after requesting.

**Comment by ara4n** [comment]
okay, so this is excellent; thanks. so twisted receives the request almost instantly (well, within 400ms or so) and processes it within 24ms... but then it takes ~4s to actually send the response (including headers) back to curl.  Presumably twisted is thrashing around doing other stuff that means it's failing to get to even send a trivial response like this.  From the fragment of logs here, GET-233078 looks a bit suspicious that curbaf (Matrix HQ) has 1696 conflicted events which it takes ~400ms to resolve, but presumably there may be other longlived federation requests churning away chewing resources and slowing everything down.

Do you have a grafana set up on this?  I'd be interested to see what the `Non-sync avg response time` graph looks like:

`rate(synapse_http_server_response_time:total{instance="$instance",job=~"$job",index=~"$index",tag!="incremental_sync"}[$bucket_size])/rate(synapse_http_server_response_time:count{instance="$instance",job=~"$job",index=~"$index",tag!="incremental_sync"}[$bucket_size])`

with a legend of

`{{job}}-{{index}} {{method}} {{servlet}} {{tag}}`

as my guess is that there are a tonne of really long lived CPU-intensive federation requests flying around making everything grind to a halt.

I'll take a look at the pyflame which might help clarify things.

**Comment by ara4n** [comment]
so the pyflame does clarify things a lot.  Ftr you turn the prof output into an SVG using flamegraph.pl:
`cat ~/Downloads/prof.txt | ./flamegraph.pl > hs.svg`, which produces https://matrix.org/~matthew/hs.svg (scroll to the bottom for the good stuff).  The interesting bits look to be:

![flame1](https://user-images.githubusercontent.com/1294269/36641300-68564546-1a25-11e8-99a0-c4c3c28bd513.png)

...where we're spending 339 samples (so 3.39s, if it's managing to hit your target sample rate of -r0.01) in _get_state_for_groups()

and then:

![flame2](https://user-images.githubusercontent.com/1294269/36641299-683b4f7a-1a25-11e8-9857-3995d1c9df51.png)

...where we're managing (coincidentally?) to spend another 339 samples in _get_state_groups_from_group_txn().

@erikjohnston will be able to interpret this better than me, but it looks to me as if we're wedging the whole reactor for ~3s whilst loading state from the DB, which would explain why trivial requests get blocked in this manner.

So, some questions:
 * I see from your config that  you have `event_cache_size: "10000K"`.  10M events is absolutely massive; we only have 200K on matrix.org.  How much RAM does the box have?  I suspect this is the problem; the cache is so big that it's causing way more problems than it solves.
 * Do you have a SYNAPSE_CACHE_FACTOR env var being set?
 * Is this postgres or sqlite? ironically I can't tell from your logs.

**Comment by Half-Shot** [comment]
Thanks for all your help @ara4n, this is certainly something that would haven taken me weeks to work out.

>I see from your config that you have event_cache_size: "10000K". 10M events is absolutely massive; event_cache_size is only 200K on matrix.org(!) How much RAM does the box have? I suspect this is the problem; the cache is so big that it's causing way more problems than it solves.

It's got about 16GB and it's an exclusively synapse box, and the owner has pretty much said "take what you need, so if you need more than 16GB then just up it.". So I figured I might as well up the cache size to something I'd never hit on the basis that it's just spare memory sitting around. But if that is actually causing more problems, then that could be it.

>Do you have a SYNAPSE_CACHE_FACTOR env var being set?

I did that migration a while back and I don't think I ever set a value, so it would be at the default. I could easily increase it if needed.

>Is this postgres or sqlite? ironically I can't tell from your logs.

Thankfully, postgres.

**Comment by erikjohnston** [comment]
> @erikjohnston will be able to interpret this better than me, but it looks to me as if we're wedging the whole reactor for ~3s whilst loading state from the DB, which would explain why trivial requests get blocked in this manner.

Firstly, that's happening in a different thread to the reactor, so it won't wedge it. Secondly, it's unlikely this is a single call and may be many small calls, so probably wouldn't wedge the reactor. 

Though spending that much time loading state is going to slow things down.

This is fixed by simply upping the cache factor, as that's what controls the size of the state cache

**Comment by Half-Shot** [comment]
I've found that restarting it and lowering the ``event_cache_size`` has made a huge difference. Though I will wait to give it a few hours of usage before I draw conclusions. It is looking good though. I'd be interested to know why this would cause a speed improvement.

**Comment by ara4n** [comment]
So it sounds like you were blowing out your state cache (not to be confused with your general-purpose event_cache_size), and increasing your synapse_cache_factor will have resized the state cache, so \o/.

I'm interested in understanding why trivial things like /versions were taking 5 seconds though, as it's similar to a problem we were seeing on overloaded sync workers last week.

@erikjohnston has some experimental grafana graphs tracking reactor health, which look like this:

<img width="1920" alt="screen shot 2018-02-25 at 14 00 00" src="https://user-images.githubusercontent.com/1294269/36642235-72a32ab4-1a34-11e8-85d4-e7777d7f9e59.png">

It might be interesting to see what these look like at both the new and old cache sizing...

`rate(python_twisted_reactor_tick_time:total{job="$job",index=~"$index",instance="$instance"}[$bucket_size])/rate(python_twisted_reactor_tick_time:count{job="$job",index=~"$index",instance="$instance"}[$bucket_size])`

and

`rate(python_twisted_reactor_pending_calls:total{job="$job",index=~"$index",instance="$instance"}[$bucket_size])/rate(python_twisted_reactor_pending_calls:count{job="$job",index=~"$index",instance="$instance"}[$bucket_size])`

**Comment by Half-Shot** [comment]
Going to go ahead and close this since my issue is now fixed.
TL;DR -

* Lower event_cache_size to something sensible (10k)
* Increase SYNAPSE_CACHE_FACTOR  to something large (10).
